tests/test_client/abstract_client.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use log::error;
+use parsec_client::auth::AuthenticationData;
+use parsec_client::core::basic_client::BasicClient;
+use parsec_client::error::Result;
+use parsec_interface::operations::list_providers::ProviderInfo;
+use parsec_interface::operations::psa_algorithm::{Algorithm, AsymmetricSignature, Hash};
+use parsec_interface::operations::psa_key_attributes::*;
+use parsec_interface::operations::psa_key_attributes::{KeyAttributes, KeyPolicy, UsageFlags};
+use parsec_interface::requests::{Opcode, ProviderID};
+use std::collections::{HashMap, HashSet};
+
+/// Client structure automatically choosing a provider and high-level operation functions.
+#[derive(Debug)]
+pub struct TestClient {
+    op_client: BasicClient,
+    cached_opcodes: Option<HashMap<ProviderID, HashSet<Opcode>>>,
+    provider: Option<ProviderID>,
```

Comment:
Could this be the same provider as the one in the basic client?

Reply:
I was thinking about that, but given how the whole provider discovery thing works, this is the way to force a test to use one specific provider, because the flow goes like this:
* you ask for an operation to be executed, say, generate key
* the client checks to see if you hardcoded a value (this `Option<ProviderID>`)
* if yes, it uses that
* if no, it checks the cache for a provider that it can use

Most of our tests are provider-independent so would have to include some sort of "active provider" discovery mechanism. But it's also useful to have the option to force the client to use a particular provider. This can't be done with the `implicit_provider` of the underlying `BasicClient` because that one isn't optional, so you wouldn't know whether it was hardcoded by the test or if it was there from the previous request.

Reply:
Also, I just realised - do we actually have any tests where we attempt to make a request against a provider that isn't active?

Reply:
> this is the way to force a test to use one specific provider

I do not think we are actually using that feature! The only occurence of `set_provider` in our tests is in `basic::wrong_provider_code` to test that a crypto operation on core fails. That wouldn't be possible at the `BasicClient` level anyway because of the new restrictions.

That could be useful in the `all_providers` tests, if you want to execute requests on multiple providers.

> This can't be done with the implicit_provider of the underlying BasicClient because that one isn't optional, so you wouldn't know whether it was hardcoded by the test or if it was there from the previous request.

Ok I think I get you, you can change the provider ID just for one request and then set it to `None` again so that it uses the one originally selected by the discovery.

Can you do that with a method `replace_provider` that replace the current `implicit_provider` with the one given and returns the old one? That way this logic could be done in the test itself when and if it is needed in the future.

The reason I am asking is that if we remove this logic, we can actually use in all of our tests the `BasicClient` methods directly for most operations and not have this wrapper in the test client. That would make the code much simpler I think.

Reply:
> Also, I just realised - do we actually have any tests where we attempt to make a request against a provider that isn't active?

I don't think so, and it would be good to add!